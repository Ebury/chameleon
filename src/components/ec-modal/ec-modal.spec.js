/* eslint-disable no-use-before-define */
import { mount } from '@vue/test-utils';
import EcModal from './ec-modal.vue';

jest.mock('../../directives/ec-focus-trap');

function mountModal(props, mountOpts) {
  return mount(EcModal, {
    propsData: { ...props },
    ...mountOpts,
  });
}

describe('EcModal', () => {
  it('should not render the modal if "showModal" is not set to true', () => {
    const wrapper = mountModal();
    expect(wrapper.find('.ec-modal').exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render basic modal', () => {
    const wrapper = mountModal(
      { showModal: true },
      {
        slots: {
          main: '<p>Before we can process your application we need you to upload your management accounts. You can do this now or leave it for later.</p>',
        },
      },
    );

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render close button', () => {
    const wrapper = mountModal({ showModal: true });

    expect(wrapper.find('.ec-modal__close').exists()).toBe(true);
    expect(wrapper.find('.ec-modal__header').element).toMatchSnapshot();
  });

  it('should not render close button when isClosable is set to false', () => {
    const wrapper = mountModal({
      showModal: true,
      isClosable: false,
    });

    expect(wrapper.find('.ec-modal__close').exists()).toBe(false);
    expect(wrapper.find('.ec-modal__header').element).toMatchSnapshot();
  });

  it('should have the ec-modal--lg class', () => {
    const wrapper = mountModal({
      showModal: true,
      large: true,
    });

    expect(wrapper.find('.ec-modal__content').classes('ec-modal--lg')).toBe(true);
    expect(wrapper.find('.ec-modal__content').element).toMatchSnapshot();
  });

  it('should not render footer left section if slot not passed', () => {
    const wrapper = mountModal({
      showModal: true,
    });

    expect(wrapper.find('.ec-modal__footer-left-content').exists()).toBe(false);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should render footer left section when slot is passed', () => {
    const wrapper = mountModal({
      showModal: true,
    },
    {
      slots: {
        footerLeftContent: '<p>Need Help ?</p>',
      },
    });

    expect(wrapper.find('.ec-modal__footer-left-content').exists()).toBe(true);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should render negative button if slot is passed', () => {
    const wrapper = mountModal({
      showModal: true,
    },
    {
      slots: {
        negative: 'Skip for now',
      },
    });

    expect(wrapper.find('.ec-modal__negative-btn').exists()).toBe(true);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should not render negative button if slot not passed', () => {
    const wrapper = mountModal({
      showModal: true,
    });

    expect(wrapper.find('.ec-modal__negative-btn').exists()).toBe(false);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should render positive button if slot is passed', () => {
    const wrapper = mountModal({
      showModal: true,
    },
    {
      slots: {
        positive: 'Update management accounts',
      },
    });

    expect(wrapper.find('.ec-modal__positive-btn').exists()).toBe(true);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should not render positive button if slot not passed', () => {
    const wrapper = mountModal({
      showModal: true,
    });

    expect(wrapper.find('.ec-modal__positive-btn').exists()).toBe(false);
    expect(wrapper.find('.ec-modal__footer').element).toMatchSnapshot();
  });

  it('should emit a "positive" event when clicking on the positive button', () => {
    const wrapper = mountModal(
      {
        showModal: true,
      },
      {
        slots: {
          positive: 'Positive Button',
        },
      },
    );
    wrapper.find('.ec-modal__positive-btn').trigger('click');
    expect(wrapper.emitted().positive).toBeTruthy();
  });

  it('should emit a "negative" event when clicking on the negative button', () => {
    const wrapper = mountModal({
      showModal: true,
    },
    {
      slots: {
        negative: 'Negative Button',
      },
    });

    wrapper.find('.ec-modal__negative-btn').trigger('click');
    expect(wrapper.emitted().negative).toBeTruthy();
  });

  it('should emit a "close" event when clicking on the close button', () => {
    const wrapper = mountModal({
      showModal: true,
      isClosable: true,
    });

    wrapper.find('.ec-modal__close').trigger('click');
    expect(wrapper.emitted().close).toBeTruthy();
  });

  it('should close the modal if ESC key is pressed and is closable', () => {
    const wrapper = mountModal({
      showModal: true,
      isClosable: true,
    }, {
      attachToDocument: true,
    });

    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close).toBeTruthy();

    wrapper.destroy(); // because we attached the wrapper to document
  });

  it('should not close the modal if ESC key is pressed and is not closable', () => {
    const wrapper = mountModal({
      showModal: true,
      isClosable: false,
    }, {
      attachToDocument: true,
    });

    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close).toBeUndefined();

    wrapper.destroy(); // because we attached the wrapper to document
  });

  it('should not close the modal if key other than ESC is pressed and is closable', () => {
    const wrapper = mountModal({
      showModal: true,
      isClosable: true,
    }, {
      attachToDocument: true,
    });

    wrapper.trigger('keyup', { key: 'a' });
    expect(wrapper.emitted().close).toBeUndefined();

    wrapper.destroy(); // because we attached the wrapper to document
  });

  it('should stop listening to keyup events when closed', () => {
    const addEventListenerSpy = jest.spyOn(document, 'addEventListener');
    const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener');

    expect(addEventListenerSpy).not.toHaveBeenCalled();

    const wrapper = mountModal({
      showModal: true,
      isClosable: false,
    }, {
      attachToDocument: true,
    });

    expect(addEventListenerSpy).toHaveBeenCalledTimes(1);
    expect(addEventListenerSpy).toHaveBeenCalledWith('keyup', expect.any(Function));

    wrapper.setProps({ showModal: false });

    expect(addEventListenerSpy).toHaveBeenCalledTimes(1);
    expect(removeEventListenerSpy).toHaveBeenCalledWith('keyup', addEventListenerSpy.mock.calls[0][1]);

    wrapper.destroy(); // because we attached the wrapper to document
  });
});
