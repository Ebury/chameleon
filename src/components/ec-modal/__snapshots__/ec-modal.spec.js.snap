// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcModal should disable the negative button when the "isDisabled" given prop is true 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <button
    class="ec-modal__negative-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--secondary ec-btn--disabled"
    data-test="ec-modal__negative-btn ec-btn"
    disabled=""
    type="button"
  >
    <!--v-if-->
    <span
      class="ec-btn__text"
      data-test="ec-btn__text"
    >
      
      
      negative Button
      
      
    </span>
    <!--v-if-->
  </button>,
}
`;

exports[`EcModal should disable the positive button when the "isDisabled" given prop is true 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <button
    class="ec-modal__positive-btn--right ec-modal__positive-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--primary ec-btn--disabled"
    data-test="ec-modal__positive-btn ec-btn"
    disabled=""
    type="button"
  >
    <!--v-if-->
    <span
      class="ec-btn__text"
      data-test="ec-btn__text"
    >
      
      
      Positive Button
      
      
    </span>
    <!--v-if-->
  </button>,
}
`;

exports[`EcModal should have the ec-modal--lg class 1`] = `
<div
  class="ec-modal__content ec-modal--lg"
  data-test="ec-modal__content"
>
  <header
    class="ec-modal__header"
    data-test="ec-modal__header"
  >
    
    
    <a
      class="ec-modal__close"
      data-test="ec-modal__close"
      href="#"
    >
      <svg
        class="ec-icon ec-modal__close-icon"
        data-test="ec-modal__close-icon"
        height="24"
        width="24"
      >
        <use
          xlink:href="#ec-simple-close"
        />
      </svg>
    </a>
  </header>
  <main
    class="ec-modal__main"
    data-test="ec-modal__main"
  >
    
    
  </main>
  <!--v-if-->
</div>
`;

exports[`EcModal should have the style attribute z-index when the z-index props is given 1`] = `
<div
  class="ec-modal"
  data-test="ec-modal"
  style="z-index: 235;"
>
  <div
    class="ec-modal__content"
    data-test="ec-modal__content"
  >
    <header
      class="ec-modal__header"
      data-test="ec-modal__header"
    >
      
      
      <a
        class="ec-modal__close"
        data-test="ec-modal__close"
        href="#"
      >
        <svg
          class="ec-icon ec-modal__close-icon"
          data-test="ec-modal__close-icon"
          height="24"
          width="24"
        >
          <use
            xlink:href="#ec-simple-close"
          />
        </svg>
      </a>
    </header>
    <main
      class="ec-modal__main"
      data-test="ec-modal__main"
    >
      
      
    </main>
    <!--v-if-->
  </div>
</div>
`;

exports[`EcModal should not render close button when isClosable is set to false 1`] = `
<header
  class="ec-modal__header"
  data-test="ec-modal__header"
>
  
  
  <!--v-if-->
</header>
`;

exports[`EcModal should not render the modal if "show" is not set to true 1`] = `
<transition-stub>
  <!--v-if-->
</transition-stub>
`;

exports[`EcModal should render basic modal 1`] = `
<transition-stub>
  <div
    class="ec-modal"
    data-test="ec-modal"
  >
    <div
      class="ec-modal__content"
      data-test="ec-modal__content"
    >
      <header
        class="ec-modal__header"
        data-test="ec-modal__header"
      >
        
        
        <a
          class="ec-modal__close"
          data-test="ec-modal__close"
          href="#"
        >
          <svg
            class="ec-icon ec-modal__close-icon"
            data-test="ec-modal__close-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#ec-simple-close"
            />
          </svg>
        </a>
      </header>
      <main
        class="ec-modal__main"
        data-test="ec-modal__main"
      >
        
        <p>
          Before we can process your application we need you to upload your management accounts. You can do this now or leave it for later.
        </p>
        
      </main>
      <!--v-if-->
    </div>
  </div>
</transition-stub>
`;

exports[`EcModal should render close button 1`] = `
<header
  class="ec-modal__header"
  data-test="ec-modal__header"
>
  
  
  <a
    class="ec-modal__close"
    data-test="ec-modal__close"
    href="#"
  >
    <svg
      class="ec-icon ec-modal__close-icon"
      data-test="ec-modal__close-icon"
      height="24"
      width="24"
    >
      <use
        xlink:href="#ec-simple-close"
      />
    </svg>
  </a>
</header>
`;

exports[`EcModal should render footer left section when slot is passed 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <div
    class="ec-modal__footer-left-content"
    data-test="ec-modal__footer-left-content"
  >
    
    <p>
      Need Help ?
    </p>
    
  </div>
  <!--v-if-->
  <!--v-if-->
</footer>
`;

exports[`EcModal should render negative button if slot is passed 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading ec-modal__btn-loading--right"
    data-test="ec-loading"
  >
    <!--v-if-->
    <div
      class="ec-loading__content ec-loading__content--is-transparent"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__negative-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--secondary"
        data-test="ec-modal__negative-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Skip for now
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
  <!--v-if-->
</footer>
`;

exports[`EcModal should render negative button with given category 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading ec-modal__btn-loading--right"
    data-test="ec-loading"
  >
    <!--v-if-->
    <div
      class="ec-loading__content ec-loading__content--is-transparent"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__negative-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--warning"
        data-test="ec-modal__negative-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Negative button
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
  <!--v-if-->
</footer>
`;

exports[`EcModal should render negative button with the loading status if slot is passed 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading ec-modal__btn-loading--right"
    data-test="ec-loading"
  >
    <div
      class="ec-loading__backdrop"
      data-test="ec-loading__backdrop"
    >
      <svg
        class="ec-icon ec-loading__icon"
        data-test="ec-loading__icon"
        height="48"
        width="48"
      >
        <use
          xlink:href="#ec-simple-loading"
        />
      </svg>
    </div>
    <div
      class="ec-loading__content"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__negative-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--secondary"
        data-test="ec-modal__negative-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Skip for now
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
  <!--v-if-->
</footer>
`;

exports[`EcModal should render positive button if slot is passed 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading"
    data-test="ec-loading"
  >
    <!--v-if-->
    <div
      class="ec-loading__content ec-loading__content--is-transparent"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__positive-btn--right ec-modal__positive-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--primary"
        data-test="ec-modal__positive-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Update management accounts
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
</footer>
`;

exports[`EcModal should render positive button with given category 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading"
    data-test="ec-loading"
  >
    <!--v-if-->
    <div
      class="ec-loading__content ec-loading__content--is-transparent"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__positive-btn--right ec-modal__positive-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--error"
        data-test="ec-modal__positive-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Positive button
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
</footer>
`;

exports[`EcModal should render positive button with the loading status if slot is passed 1`] = `
<footer
  class="ec-modal__footer"
  data-test="ec-modal__footer"
>
  <!--v-if-->
  <!--v-if-->
  <div
    class="ec-loading ec-modal__btn-loading"
    data-test="ec-loading"
  >
    <div
      class="ec-loading__backdrop"
      data-test="ec-loading__backdrop"
    >
      <svg
        class="ec-icon ec-loading__icon"
        data-test="ec-loading__icon"
        height="48"
        width="48"
      >
        <use
          xlink:href="#ec-simple-loading"
        />
      </svg>
    </div>
    <div
      class="ec-loading__content"
      data-test="ec-loading__content"
    >
      
      <button
        class="ec-modal__positive-btn--right ec-modal__positive-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--primary"
        data-test="ec-modal__positive-btn ec-btn"
        type="button"
      >
        <!--v-if-->
        <span
          class="ec-btn__text"
          data-test="ec-btn__text"
        >
          
          
          Update management accounts
          
          
        </span>
        <!--v-if-->
      </button>
      
    </div>
  </div>
</footer>
`;

exports[`EcModal should render with an enabled negative button by default 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <button
    class="ec-modal__negative-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--secondary"
    data-test="ec-modal__negative-btn ec-btn"
    type="button"
  >
    <!--v-if-->
    <span
      class="ec-btn__text"
      data-test="ec-btn__text"
    >
      
      
      Negative Button
      
      
    </span>
    <!--v-if-->
  </button>,
}
`;

exports[`EcModal should render with an enabled positive button by default 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <button
    class="ec-modal__positive-btn--right ec-modal__positive-btn ec-btn ec-btn--md ec-btn--rounded ec-btn--primary"
    data-test="ec-modal__positive-btn ec-btn"
    type="button"
  >
    <!--v-if-->
    <span
      class="ec-btn__text"
      data-test="ec-btn__text"
    >
      
      
      Positive Button
      
      
    </span>
    <!--v-if-->
  </button>,
}
`;

exports[`EcModal v-model should not render the modal when we pass to model false 1`] = `
<transition-stub>
  <!--v-if-->
</transition-stub>
`;

exports[`EcModal v-model should render the modal when we pass to model true 1`] = `
<transition-stub>
  <div
    class="ec-modal"
    data-test="ec-modal"
  >
    <div
      class="ec-modal__content"
      data-test="ec-modal__content"
    >
      <header
        class="ec-modal__header"
        data-test="ec-modal__header"
      >
        
        
        <a
          class="ec-modal__close"
          data-test="ec-modal__close"
          href="#"
        >
          <svg
            class="ec-icon ec-modal__close-icon"
            data-test="ec-modal__close-icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="#ec-simple-close"
            />
          </svg>
        </a>
      </header>
      <main
        class="ec-modal__main"
        data-test="ec-modal__main"
      >
        
        
      </main>
      <!--v-if-->
    </div>
  </div>
</transition-stub>
`;

exports[`EcModal v-model should render the modal when we pass to model true 2`] = `
<transition-stub>
  <!--v-if-->
</transition-stub>
`;

exports[`EcModal when second modal is present should be rendered with a second modal 1`] = `
<div
  class="ec-modal"
  data-test="ec-modal"
>
  <div
    class="ec-modal__content ec-modal--lg"
    data-test="ec-modal__content"
  >
    <header
      class="ec-modal__header"
      data-test="ec-modal__header"
    >
      
      <h2>
        First Modal
      </h2>
      
      <a
        class="ec-modal__close"
        data-test="ec-modal__close"
        href="#"
      >
        <svg
          class="ec-icon ec-modal__close-icon"
          data-test="ec-modal__close-icon"
          height="24"
          width="24"
        >
          <use
            xlink:href="#ec-simple-close"
          />
        </svg>
      </a>
    </header>
    <main
      class="ec-modal__main"
      data-test="ec-modal__main"
    >
      
      <div>
        <button
          data-test="first-modal-action__btn"
        >
          Action first modal
        </button>
        <p
          class="tw-mt-0"
        >
          Before we can process your application we need you to upload your management accounts. You can do this now or leave it for later.
        </p>
      </div>
      
    </main>
    <!--v-if-->
  </div>
</div>
`;
